File: src/main/java/org/jetbrains/kotlin/resolve/lang/java/PackageSearchers.java
/**
 * *****************************************************************************
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************************
 */
package org.jetbrains.kotlin.resolve.lang.java;

import com.google.common.collect.Lists;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import javax.lang.model.element.Element;
import javax.lang.model.element.PackageElement;
import javax.lang.model.element.TypeElement;
import kotlin.jvm.functions.Function1;
import org.jetbrains.kotlin.load.java.structure.JavaClass;
import org.jetbrains.kotlin.load.java.structure.JavaPackage;
import org.jetbrains.kotlin.name.FqName;
import org.jetbrains.kotlin.name.Name;
import org.jetbrains.kotlin.resolve.lang.java.structure.NetBeansJavaClass;
import org.jetbrains.kotlin.resolve.lang.java.structure.NetBeansJavaPackage;
import org.netbeans.api.java.source.CompilationController;
import org.netbeans.api.java.source.JavaSource.Phase;
import org.netbeans.api.java.source.Task;
import org.netbeans.api.project.Project;

/**
 *
 * @author Alexander.Baratynski
 */
public class PackageSearchers {

    public static class SubPackagesSearcher implements Task<CompilationController> {

        private final JavaPackage pack;
        private final Project project;
        private final Collection<JavaPackage> subPackages = Lists.newArrayList();

        public SubPackagesSearcher(Project project, JavaPackage pack) {
            this.pack = pack;
            this.project = project;
        }

        public PackageElement[] findPackageFragments(String name,
                boolean partialMatch, boolean patternMatch, CompilationController info) {
            Set<String> packages = NBElementUtils.getPackages(project, name);
            List<PackageElement> subpackageElements = Lists.newArrayList();
            for (String pack : packages) {
                PackageElement subpackageElement = info.getElements().getPackageElement(name);
                if (subpackageElement == null) {
                    continue;
                }
                subpackageElements.add(subpackageElement);
            }

            if (subpackageElements.isEmpty()) {
                return null;
            }

            return subpackageElements.toArray(new PackageElement[subpackageElements.size()]);
        }

        @Override
        public void run(CompilationController info) throws Exception {
            info.toPhase(Phase.RESOLVED);
            String thisPackageName = pack.getFqName().asString();
            String pattern = thisPackageName.isEmpty() ? "*" : thisPackageName + ".";

            PackageElement[] packageFragments = findPackageFragments(pattern, true, true, info);

            int thisNestedLevel = thisPackageName.split("\\.").length;
            if (packageFragments != null && packageFragments.length > 0) {
                for (PackageElement packageFragment : packageFragments) {
                    int subNestedLevel = packageFragment.getQualifiedName().toString().split("\\.").length;
                    boolean applicableForRootPackage = thisNestedLevel == 1 && thisNestedLevel == subNestedLevel;
                    if (!packageFragment.getQualifiedName().toString().isEmpty()
                            && (applicableForRootPackage || (thisNestedLevel + 1 == subNestedLevel))) {
                        subPackages.add(new NetBeansJavaPackage(ElemHandle.create(packageFragment, project), project));
                    }
                }
            }
        }

        public Collection<JavaPackage> getSubPackages() {
            return subPackages;
        }

    }

    public static class ClassesSearcher implements Task<CompilationController> {

        private final Project project;
        private final List<ElemHandle<PackageElement>> packages;
        private final Function1<? super Name, Boolean> nameFilter;
        private final Collection<JavaClass> classes = Lists.newArrayList();

        public ClassesSearcher(List<ElemHandle<PackageElement>> packages, Project project, 
                Function1<? super Name, Boolean> nameFilter) {
            this.packages = packages;
            this.project = project;
            this.nameFilter = nameFilter;
        }

        private boolean isOuterClass(TypeElement classFile){
            return !classFile.getSimpleName().toString().contains("$");
        }
        
        private List<JavaClass> getClassesInPackage(PackageElement javaPackage,
                Function1<? super Name, ? extends Boolean> nameFilter) {
            List<JavaClass> javaClasses = Lists.newArrayList();
            List<? extends Element> classes = javaPackage.getEnclosedElements();

            for (Element cl : classes) {
                if (isOuterClass((TypeElement) cl)) {
                    String elementName = cl.getSimpleName().toString();
                    if (Name.isValidIdentifier(elementName) && nameFilter.invoke(Name.identifier(elementName))) {
                        javaClasses.add(new NetBeansJavaClass(ElemHandle.create(cl, project), project));
                    }
                }
            }

            return javaClasses;
        }

        @Override
        public void run(CompilationController info) throws Exception {
            info.toPhase(Phase.RESOLVED);
            
            for (ElemHandle<PackageElement> packHandle : packages) {
                Element elem = packHandle.resolve(info);
                if (elem == null) {
                    continue;
                }
                
                classes.addAll(getClassesInPackage((PackageElement) elem, nameFilter));
            }
            
        }

        public Collection<JavaClass> getClasses() {
            return classes;
        }

    }
    
    public static class FqNameSearcher implements Task<CompilationController> {

        private final ElemHandle<PackageElement> handle;
        private FqName fqName = null;
        
        public FqNameSearcher(ElemHandle<PackageElement> handle) {
            this.handle = handle;
        }
        
        @Override
        public void run(CompilationController info) throws Exception {
            info.toPhase(Phase.RESOLVED);
            Element pack = handle.resolve(info);
            if (pack == null) {
                return;
            }
            
            fqName = new FqName(((PackageElement) pack).getQualifiedName().toString());
        }
        
        public FqName getFqName() {
            return fqName;
        }
    }

}

File: src/main/java/org/jetbrains/kotlin/resolve/lang/java/packageSearchers.kt
/*******************************************************************************
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *******************************************************************************/
package org.jetbrains.kotlin.resolve.lang.java

import javax.lang.model.element.Element
import javax.lang.model.element.PackageElement
import javax.lang.model.element.TypeElement
import org.jetbrains.kotlin.load.java.structure.JavaClass
import org.jetbrains.kotlin.load.java.structure.JavaPackage
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.name.Name
import org.jetbrains.kotlin.resolve.lang.java.structure.NetBeansJavaClass
import org.jetbrains.kotlin.resolve.lang.java.structure.NetBeansJavaPackage
import org.netbeans.api.java.source.CompilationController
import org.netbeans.api.java.source.JavaSource.Phase
import org.netbeans.api.java.source.Task
import org.netbeans.api.project.Project

class SubPackagesSearcher(val project: Project,
                          val pack: JavaPackage) : Task<CompilationController> {

    val subPackages = arrayListOf<JavaPackage>()

    fun findPackageFragments(name: String, partialMatch: Boolean,
                             patternMatch: Boolean, info: CompilationController): Array<PackageElement>? {
        val packages = name.getPackages(project)
        val subpackageElements = packages.map { info.elements.getPackageElement(it) }.filterNotNull()
        if (subpackageElements.isEmpty()) return null

        return subpackageElements.toTypedArray()
    }

    override fun run(info: CompilationController) {
        info.toResolvedPhase()

        val thisPackageName = pack.fqName.asString()
        val pattern = if (thisPackageName.isEmpty()) "*" else thisPackageName + "."

        val packageFragments = findPackageFragments(pattern, true, true, info)
        val thisNestedLevel = thisPackageName.split("\\.").size

        if (packageFragments != null && packageFragments.size > 0) {
            packageFragments.forEach {
                val subNestedLevel = it.qualifiedName.toString().split("\\.").size
                val applicableForRootPackage = thisNestedLevel == 1 && thisNestedLevel == subNestedLevel
                if (!it.qualifiedName.toString().isEmpty() &&
                        (applicableForRootPackage || (thisNestedLevel + 1 == subNestedLevel))) {
                    subPackages.add(NetBeansJavaPackage(ElemHandle.create(it, project), project))
                }
            }
        }
    }
}

class ClassesSearcher(val packages: List<ElemHandle<PackageElement>>,
                      val project: Project,
                      val nameFilter: (Name) -> Boolean) : Task<CompilationController> {

    val classes = arrayListOf<JavaClass>()

    fun isOuterClass(classFile: TypeElement) = !classFile.simpleName.toString().contains("$")

    fun getClassesInPackage(javaPackage: PackageElement, nameFilter: (Name) -> Boolean): List<JavaClass> {
        return javaPackage.enclosedElements
                .filter { isOuterClass(it as TypeElement) }
                .filter {
                    Name.isValidIdentifier(it.simpleName.toString())
                            && nameFilter.invoke(Name.identifier(it.simpleName.toString()))
                }
                .map { NetBeansJavaClass(ElemHandle.create(it as TypeElement, project), project) }
    }

    override fun run(info: CompilationController) {
        info.toResolvedPhase()

        for (it in packages) {
            val elem = it.resolve(info) ?: continue
            classes.addAll(getClassesInPackage(elem as PackageElement, nameFilter))
        }
    }
}

class FqNameSearcher(val handle: ElemHandle<PackageElement>) : Task<CompilationController> {
    
    lateinit var fqName: FqName
    
    override fun run(info: CompilationController) {
        info.toResolvedPhase()
        
        val pack = handle.resolve(info) ?: throw UnsupportedOperationException("Couldn''t resolve $handle")
        fqName = FqName((pack as PackageElement).qualifiedName.toString())
    }
}










